src/ConnectionPool.jl

```diff
@@ -483,59 +473,87 @@ function getconnection(::Type{Transaction{T}},
-                                                  kw...)
-                c = Connection(host, port,
-                               pipeline_limit, idle_timeout,
-                               require_ssl_verification,
-                               io)
-                push!(pool[Threads.threadid()], c)                ;@debug 1 "🔗  New:            $c"
-                return client_transaction(c)
-            end
+    # Wait for `closewrite` or `close` to signal that a connection is ready.
+    # sym = gensym()
+    # @debug 0 "$sym: connection pool full, waiting..."
+    # tt = time()
+    while true
+        conn = take!(pod.conns)
+        if isvalid(pod, conn, reuse_limit, pipeline_limit)
+            # this is a reuseable connection, so use it
+            @debug 2 "2 reusing connection: $conn"
+            return client_transaction(conn)
+        elseif pod.numactive[] < connection_limit
+            return newconnection(pod, T, host, port, pipeline_limit,
+                require_ssl_verification, idle_timeout; kw...)
+        end
+    end
+end
 
-            # Share a connection that has active readers...
-            if !isempty(writable)
-                c = rand(writable)             ;@debug 2 "⇆  Shared:         $c"
-                return client_transaction(c)
+function isvalid(pod, conn, reuse_limit, pipeline_limit)
+    # Close connections that have reached the reuse limit...
+    if reuse_limit != nolimit
+        if conn.readcount[] >= reuse_limit && !readbusy(conn)
+            @debug 2 "💀 overuse:         $conn"
+            close(conn.io)
+        end
+    end
+    # Close connections that have reached the timeout limit...
+    if conn.idle_timeout > 0
+        if !isbusy(conn) && inactiveseconds(conn) > conn.idle_timeout
+            @debug 2 "💀 idle timeout:    $conn"
+            close(conn.io)
+        end
+    end
+    # For closed connections, we decrease active count in pod, and "continue"
+    # which effectively drops the connection
+    if !isopen(conn.io)
+        close(conn)
+        lock(conn.closelock) do
+            if !conn.closed
+                conn.closed = true
+                decr!(pod)
             end
-        finally
-            unlock(poollock[Threads.threadid()])
         end
-
-        # Wait for `closewrite` or `close` to signal that a connection is ready.
-        wait(poolcondition[Threads.threadid()])
+        return false
+    end
+    # If we've hit our pipeline_limit, can't use this one, but don't close
+    if (conn.writecount[] - conn.readcount[]) >= pipeline_limit + 1
+        return false
     end
+
+    return !writebusy(conn)
+end
+
+function newconnection(pod, T, host, port, pipeline_limit, require_ssl_verification, idle_timeout; kw...)
+    io = getconnection(T, host, port;
+            require_ssl_verification=require_ssl_verification, kw...)
+    c = Connection(host, port, pipeline_limit, idle_timeout, require_ssl_verification, io)
+    incr!(pod)
```

Comment:
This is not atomic without something also being synchronized—probably should have a lock and should not use `Atomic`